lib/spack/spack/solver/concretize.lp

```diff
@@ -51,10 +50,38 @@ depends_on(Package, Dependency, Type)
   depends_on(Package, Provider, Type)
   : provides_virtual(Provider, Virtual)
 } 1
- :- declared_dependency(Package, Virtual, Type),
+ :- dependency_conditions(Package, Virtual, Type),
     virtual(Virtual),
-    not external(Package),
-    node(Package).
+    not external(Package).
+
+% if any individual condition below is true, trigger the dependency.
+dependency_conditions(P, D, T) :- dependency_conditions(P, D, T, _).
+
+% collect all the dependency condtions into a single conditional rule
+dependency_conditions(P, D, T, I) :-
+  node(Package)
+    : dep_cond(P, D, T, I, "node", Package);
+  version(Package, Version)
+    : dep_cond(P, D, T, I, "version", Package, Version);
+  version_satisfies(Package, Constraint)
+    : dep_cond(P, D, T, I, "version_satisfies", Package, Constraint);
+  node_platform(Package, Platform)
+    : dep_cond(P, D, T, I, "node_platform", Package, Platform);
+  node_os(Package, OS)
+    : dep_cond(P, D, T, I, "node_os", Package, OS);
+  node_target(Package, Target)
+    : dep_cond(P, D, T, I, "node_target", Package, Target);
+  variant_value(Package, Variant, Value)
+    : dep_cond(P, D, T, I, "variant_value", Package, Variant, Value);
+  node_compiler(Package, Compiler)
+    : dep_cond(P, D, T, I, "node_compiler", Package, Compiler);
+  node_compiler_version(Package, Compiler, Version)
+    : dep_cond(P, D, T, I, "node_compiler_version", Package, Compiler, Version);
+  node_flag(Package, FlagType, Flag)
+    : dep_cond(P, D, T, I, "node_flag", Package, FlagType, Flag);
+  dependency_condition(P, D, I);
```

Comment:
Can we rename this fact? Maybe something like:
```suggestion
  evaluate_dependency_condition(Package, Dependency, ID);
```
? I'm a bit worried about having two completely different terms that are one the plural of the other (`dependency_conditions` vs. `dependency_condition`).